File: src/main/java/com/github/bartgora/GameCharacter.java
package com.github.bartgora;

public abstract class GameCharacter {

    public String render() {
        return sayHello() + " I have a " + getWeapon() + " And I can " + getAbility();
    }

    public String sayHello() {
        return "My name is " + getName();
    }

    protected abstract String getName();

    protected abstract String getWeapon();

    protected abstract String getAbility();
}

File: src/main/kotlin/com/github/bartgora/GameCharacter.kt
package com.github.bartgora

abstract class GameCharacter {

    fun render(): String {
        return sayHello() + " I have a " + weapon + " And I can " + ability
    }

    fun sayHello(): String {
        return "My name is $name"
    }

    protected abstract val name: String
    protected abstract val weapon: String
    protected abstract val ability: String
}
